### PR TITLE
Add lazy parsing feature

### PR DESCRIPTION
**Pull request by AlexKlo**
### Pull request checklist

- [x] Did you bump the project version?
- [x] Did you add a description to the Pull Request?
- [x] Did you add new tests?
- [x] Did you add reviewers?

Implement easy lazy parsing based on yyjson interfaces.

### Regular parser
```
function yyjson(json)
    v = parse_json(json)
    return v["symbols"][1]["filters"][1]["filterType"]
end

@benchmark yyjson(json_str)
# BenchmarkTools.Trial: 136 samples with 1 evaluation.
#  Range (min … max):  28.658 ms … 59.495 ms  ┊ GC (min … max):  0.00% … 37.72%
#  Time  (median):     33.859 ms              ┊ GC (median):     0.00%
#  Time  (mean ± σ):   36.792 ms ±  7.689 ms  ┊ GC (mean ± σ):  13.09% ± 14.09%

#     █▄█                                                        
#   ▄▅███▃█▄▄▆▄▆▄▃▅▆▁▆▅▅▁▁▄▃▃▆▁▃▁▅▃▄▃▄▃▅▃▃▁▁▃▁▃▁▄▃▃▄▃▃▄▃▃▁▁▁▄▁▃ ▃
#   28.7 ms         Histogram: frequency by time        56.4 ms <

#  Memory estimate: 34.45 MiB, allocs estimate: 680374.
```

### Lazy parser
```
function yyjson_lazy(json)
    v = lazy_parse(json)
    try
        v["symbols"][1]["filters"][1]["filterType"]
    finally
        close(v)
    end
end

@benchmark yyjson_lazy(json_str)
# BenchmarkTools.Trial: 628 samples with 1 evaluation.
#  Range (min … max):  6.779 ms …  13.161 ms  ┊ GC (min … max): 0.00% … 0.00%
#  Time  (median):     7.835 ms               ┊ GC (median):    0.00%
#  Time  (mean ± σ):   7.913 ms ± 602.214 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

#           ▁  ▃▅▅▆▅█▂▁█▄▅▄▁▁▂ ▁                                 
#   ▂▃▃▃▃▆▆▇████████████████████▄▆▃▃▆▂▃▃▃▂▂▃▃▁▃▁▂▂▂▂▂▁▁▁▁▁▁▁▁▁▂ ▄
#   6.78 ms         Histogram: frequency by time        10.3 ms <

#  Memory estimate: 352 bytes, allocs estimate: 9.
```

```
julia> versioninfo()
Julia Version 1.11.1
Commit 8f5b7ca12ad (2024-10-16 10:53 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 20 × 12th Gen Intel(R) Core(TM) i7-12700H
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, alderlake)
Threads: 1 default, 0 interactive, 1 GC (on 20 virtual cores)
```

### CODE REVIEW

File: src/LazyParser.jl [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+end
+
+function Base.iterate(obj::LazyDict)
+    iter = obj.iter
+    iter_ptr = pointer_from_objref(iter)
+    GC.@preserve iter begin
+        yyjson_obj_iter_init(obj.ptr, iter_ptr) || throw(LazyYYJSONError("Failed to initialize object iterator"))
+
+        yyjson_obj_iter_has_next(iter_ptr) || return nothing
+
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
```

File: src/LazyParser.jl [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+#__ LazyDict
+struct LazyDict{T<:Ptr{YYJSONVal}} <: AbstractDict{AbstractString, Any}
+    ptr::T
+    iter::YYJSONObjIter
+
+    function LazyDict(ptr::Ptr{YYJSONVal})
+        iter = YYJSONObjIter()
+        new{Ptr{YYJSONVal}}(ptr, iter)
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
struct LazyDict{T<:Ptr{YYJSONVal}} <: AbstractDict{AbstractString,Any}
    ptr::T
    iter::YYJSONObjIter

    function LazyDict(ptr::Ptr{YYJSONVal})
        iter = YYJSONObjIter()
        return new{Ptr{YYJSONVal}}(ptr, iter)
    end
end
```

File: src/LazyParser.jl [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+function Base.getindex(obj::LazyDict, key::String)
+    value = get(obj, key, :NOT_SET)
+    if value === :NOT_SET
+        throw(KeyError(key))
+    else
+        return value
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function Base.getindex(obj::LazyDict, key::AbstractString)
    value = get(obj, key, :NOT_SET)
    value === :NOT_SET && throw(KeyError(key))
    return value
end
```

File: src/LazyParser.jl [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+function Base.get(obj::LazyDict, key::String, default)
+    value_ptr = yyjson_obj_get(obj.ptr, key)
+    return if value_ptr != C_NULL
+        parse_value(value_ptr)
+    else
+        default
+    end    
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function Base.get(obj::LazyDict, key::AbstractString, default)
    value_ptr = yyjson_obj_get(obj.ptr, key)
    return value_ptr != C_NULL ? parse_value(value_ptr) : default
end
```

File: src/LazyParser.jl [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    iter_ptr = pointer_from_objref(iter)
+    GC.@preserve iter begin
+        key_ptr = yyjson_obj_iter_next(iter_ptr)
+        key = parse_string(key_ptr)
+        val_ptr = yyjson_obj_iter_get_val(key_ptr)
+        val = parse_value(val_ptr)
+        new_state = yyjson_obj_iter_has_next(iter_ptr)
+        return (key => val), new_state
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function _lazy_dict_next(iter_ptr::Ptr)
    key_ptr = yyjson_obj_iter_next(iter_ptr)
    val_ptr = yyjson_obj_iter_get_val(key_ptr)
    return (parse_string(key_ptr) => parse_value(val_ptr)), yyjson_obj_iter_has_next(iter_ptr)
end

function Base.iterate(obj::LazyDict, state = nothing)
    iter = obj.iter
    iter_ptr = pointer_from_objref(iter)
    GC.@preserve iter begin
        if state === nothing
            yyjson_obj_iter_init(obj.ptr, iter_ptr) ||
                throw(LazyYYJSONError("Failed to initialize iterator"))
        end
        return yyjson_obj_iter_has_next(iter_ptr) ? _lazy_dict_next(iter_ptr) : nothing
    end
end
```

File: src/LazyParser.jl [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+Base.size(x::LazyVector) = (length(x),)
+
+function Base.getindex(arr::LazyVector, index::Int)
+    value = get(arr, index, :NOT_SET)
+    if value === :NOT_SET
+        throw(BoundsError(arr, index))
+    else
+        return value
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function Base.getindex(arr::LazyVector, index::Integer)
    value = get(arr, index, :NOT_SET)
    value === :NOT_SET && throw(BoundsError(arr, index))
    return value
end
```

File: src/LazyParser.jl [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+function Base.get(arr::LazyVector, index::Int, default)
+    (1 <= index <= length(arr)) || return default
+    value_ptr = yyjson_arr_get(arr.ptr, index-1)
+    return if value_ptr != C_NULL
+        parse_value(value_ptr)
+    else
+        default
+    end 
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function Base.get(arr::LazyVector, index::Integer, default)
    (1 <= index <= length(arr)) || return default
    value_ptr = yyjson_arr_get(arr.ptr, index - 1)
    return value_ptr != C_NULL ? parse_value(value_ptr) : default
end
```

File: src/LazyParser.jl [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    alc_ptr::Ptr{YYJSONAlc}
+    root::LT
+    is_open::Bool
+
+    function JSONDoc(doc_ptr::Ptr{YYJSONDoc}, alc_ptr::Ptr{YYJSONAlc}, root::LT) where {LT <: Union{LazyDict, LazyVector}} 
+        doc = new{LT}(doc_ptr, alc_ptr, root, true)
+        finalizer(close, doc)
+        return doc
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
mutable struct JSONDoc{LT<:Union{LazyDict,LazyVector}}
    doc_ptr::Ptr{YYJSONDoc}
    alc_ptr::Ptr{YYJSONAlc}
    root::LT
    is_open::Bool

    function JSONDoc(doc_ptr::Ptr{YYJSONDoc}, alc_ptr::Ptr{YYJSONAlc}, root::LT) where {LT}
        doc = new{LT}(doc_ptr, alc_ptr, root, true)
        finalizer(close, doc)
        return doc
    end
end
```

File: src/LazyParser.jl [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    print(io, "JSON Document $LT")
+end
+
+function Base.close(doc::JSONDoc)
+    doc.is_open || return nothing
+    yyjson_doc_free(doc.doc_ptr)
+    yyjson_alc_dyn_free(doc.alc_ptr)
+    doc.is_open = false
+    return nothing
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
Base.show(io::IO, ::JSONDoc{LT}) where {LT} = print(io, "JSON Document $LT")
Base.close(doc::JSONDoc) = _close_doc(doc)

function _close_doc(doc::JSONDoc)
    doc.is_open || return nothing
    yyjson_doc_free(doc.doc_ptr)
    yyjson_alc_dyn_free(doc.alc_ptr)
    return doc.is_open = false
end
```

File: src/LazyParser.jl [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+function Base.getindex(doc::JSONDoc, key::Union{AbstractString, Int64})
+    value = getindex(doc.root, key)
+    return value
+end
+
+function Base.get(doc::JSONDoc, key::Union{AbstractString, Int64}, default)
+    value = get(doc.root, key, default)
+    return value
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
Base.getindex(doc::JSONDoc, key::Union{AbstractString,Integer}) = getindex(doc.root, key)
function Base.get(doc::JSONDoc, key::Union{AbstractString,Integer}, default)
    return get(doc.root, key, default)
end
```

File: src/LazyParser.jl [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    return if LT <: LazyDict
+        values(doc.root)
+    else
+        [v for v in doc.root]
+    end
+end
+
+Base.collect(doc::JSONDoc) = collect(doc.root)
+
+Base.length(doc::JSONDoc) = length(doc.root)
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion

Base.keys(doc::JSONDoc{LazyDict}) = keys(doc.root)
Base.keys(doc::JSONDoc{LazyVector}) = LinearIndices(1:length(doc))
Base.values(doc::JSONDoc{LazyDict}) = values(doc.root)
Base.values(doc::JSONDoc{LazyVector}) = [v for v in doc.root]
Base.collect(doc::JSONDoc) = collect(doc.root)
Base.length(doc::JSONDoc) = length(doc.root)

```

File: src/LazyParser.jl [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    elseif yyjson_is_bool(ptr)
+        yyjson_get_bool(ptr)
+    elseif yyjson_is_obj(ptr)
+        LazyDict(ptr)
+    elseif yyjson_is_arr(ptr)
+        LazyVector(ptr)
+    else
+        nothing
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function parse_value(ptr::Ptr{YYJSONVal})
    if yyjson_is_str(ptr) || yyjson_is_raw(ptr)
        parse_string(ptr)
    elseif yyjson_is_num(ptr)
        parse_number(ptr)
    elseif yyjson_is_bool(ptr)
        yyjson_get_bool(ptr)
    elseif yyjson_is_obj(ptr)
        LazyDict(ptr)
    elseif yyjson_is_arr(ptr)
        LazyVector(ptr)
    else
        nothing
    end
end
```

File: src/LazyParser.jl [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+
+function parse_string(ptr::Ptr{YYJSONVal})
+    ptr_char = yyjson_get_str(ptr)
+    ptr_char == C_NULL && throw(LazyYYJSONError("Error while parsing string: $ptr_char"))
+    return unsafe_string(ptr_char)
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function parse_string(ptr::Ptr{YYJSONVal})
    ptr_char = yyjson_get_str(ptr)
    ptr_char == C_NULL && throw(LazyYYJSONError("Error parsing string"))
    return unsafe_string(ptr_char)
end
```

File: src/LazyParser.jl [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+
+function parse_number(ptr::Ptr{YYJSONVal})
+    return if yyjson_is_real(ptr)
+        yyjson_get_real(ptr)
+    else
+        Int64(yyjson_get_num(ptr))
+    end
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function parse_number(ptr::Ptr{YYJSONVal})
    return yyjson_is_real(ptr) ? yyjson_get_real(ptr) : Int64(yyjson_get_num(ptr))
end
```

File: src/LazyParser.jl [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+
+function parse_root(doc_ptr::Ptr{YYJSONDoc})
+    root_ptr = yyjson_doc_get_root(doc_ptr)
+    root_ptr == C_NULL && throw(LazyYYJSONError("Error while parsing root: $root"))
+    root = parse_value(root_ptr)
+    return root
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function parse_root(doc_ptr::Ptr{YYJSONDoc})
    root_ptr = yyjson_doc_get_root(doc_ptr)
    root_ptr == C_NULL && throw(LazyYYJSONError("Error parsing root"))
    return parse_value(root_ptr)
end
```

File: src/LazyParser.jl [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+
+function lazy_parse(json::AbstractString; kw...)
+    allocator = yyjson_alc_dyn_new()
+    doc_ptr = read_json_doc(json; alc = allocator, kw...)
+    root = parse_root(doc_ptr)
+    doc = JSONDoc(doc_ptr, allocator, root)
+    return doc
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function lazy_parse(json::AbstractString; kw...)
    allocator = yyjson_alc_dyn_new()
    doc_ptr = read_json_doc(json; alc = allocator, kw...)
    return JSONDoc(doc_ptr, allocator, parse_root(doc_ptr))
end
```

File: src/LazyParser.jl [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+    end
+end
+
+function lazy_open(path::AbstractString; kw...)
+    allocator = yyjson_alc_dyn_new()
+    doc_ptr = open_json_doc(path; alc = allocator, kw...)
+    root = parse_root(doc_ptr)
+    doc = JSONDoc(doc_ptr, allocator, root)
+    return doc
+end
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function lazy_open(path::AbstractString; kw...)
    allocator = yyjson_alc_dyn_new()
    doc_ptr = open_json_doc(path; alc = allocator, kw...)
    return JSONDoc(doc_ptr, allocator, parse_root(doc_ptr))
end
```

File: src/LazyParser.jl [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+    return unsafe_string(ptr_char)
+end
+
+function parse_number(ptr::Ptr{YYJSONVal})
+    return yyjson_is_real(ptr) ? yyjson_get_real(ptr) : yyjson_get_int(ptr)
+end
+
+function parse_root(doc_ptr::Ptr{YYJSONDoc})
+    root_ptr = yyjson_doc_get_root(doc_ptr)
+    root_ptr == C_NULL && throw(LazyYYJSONError("Error parsing root"))
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
    root_ptr === C_NULL && throw(LazyYYJSONError("Error parsing root"))
```

File: src/LazyParser.jl [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+    doc_ptr = open_json_doc(path; alc = allocator, kw...)
+    root = parse_root(doc_ptr)
+    return JSONDoc(doc_ptr, allocator, root)
+end
+
+function lazy_open(io::IO; kw...)
+    return lazy_parse(read(io))
+end
+
+function lazy_open(f::Function, x...; kw...)
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
function lazy_open(f::Function, path::AbstractString; kw...)
```

File: src/LazyParser.jl [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+Base.show(io::IO, e::LazyYYJSONError) = print(io, e.message)
+
+#__ LazyDict
+struct LazyDict <: AbstractDict{AbstractString, Any}
+    ptr::Ptr{YYJSONVal}
+    iter::YYJSONObjIter
+
+    function LazyDict(ptr::Ptr{YYJSONVal})
+        iter = YYJSONObjIter()
+        new(ptr, iter)
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
        return new(ptr, iter)
```

File: src/LazyParser.jl [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+using ..YYJSON
+import ..YYJSON: read_json_doc, open_json_doc
+
+struct LazyYYJSONError <: Exception
+    message::String
+end
+
+Base.show(io::IO, e::LazyYYJSONError) = print(io, e.message)
+
+#__ LazyDict
```
Comment (gryumov) [inline src/LazyParser.jl]:
```suggestion
#__ LazyDict

```

File: src/LazyJSON.jl [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+    end
+end
+
+function Base.close(root::LazyJSONDict)
+    root.is_open || return nothing
+    yyjson_doc_free(root.doc_ptr)
+    yyjson_alc_dyn_free(root.alc_ptr)
+    root.is_open = false
+    return nothing
+end
```
Comment (gryumov) [inline src/LazyJSON.jl]:
```suggestion
function Base.close(obj::LazyJSONDict)
    obj.is_open || return nothing
    yyjson_doc_free(obj.doc_ptr)
    yyjson_alc_dyn_free(obj.alc_ptr)
    obj.is_open = false
    return nothing
end
```

File: src/LazyJSON.jl [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+        end
+        if yyjson_obj_iter_has_next(iter_ptr)
+            key_ptr = yyjson_obj_iter_next(iter_ptr)
+            val_ptr = yyjson_obj_iter_get_val(key_ptr)
+            return (parse_json_string(key_ptr) => parse_json_value(val_ptr)), true
+        else
+            return nothing
+        end
+    end
+end
```
Comment (gryumov) [inline src/LazyJSON.jl:87]:
```suggestion
function Base.iterate(obj::LazyJSONDict, state = nothing)
    iter = obj.iter
    iter_ptr = pointer_from_objref(iter)
    GC.@preserve iter begin
        if state === nothing
            yyjson_obj_iter_init(obj.ptr, iter_ptr) ||
                throw(LazyJSONError("Failed to initialize iterator."))
        end
        key_ptr = yyjson_obj_iter_next(iter_ptr)
        if key_ptr === C_NULL
            return nothing
        end
        val_ptr = yyjson_obj_iter_get_val(key_ptr)
        return (parse_json_string(key_ptr) => parse_json_value(val_ptr)), true
    end
end
```

File: src/LazyJSON.jl [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+
+function parse_json_root(doc_ptr::Ptr{YYJSONDoc})
+    root_ptr = yyjson_doc_get_root(doc_ptr)
+    root_ptr == C_NULL && throw(LazyJSONError("Error parsing root."))
+    return root_ptr
+end
+
+function parse_lazy_json(json::AbstractString; kw...)
+    allocator = yyjson_alc_dyn_new()
```
Comment (gryumov) [inline src/LazyJSON.jl:234]:
```suggestion
    allocator = yyjson_alc_dyn_new()
    if allocator === C_NULL
        throw(LazyJSONError("Failed to allocate memory for JSON parsing."))
    end
```

File: src/LazyJSON.jl [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+    doc = parse_lazy_json(x...; kw...)
+    try
+        f(doc)
+    finally
+        close(doc)
+    end
+end
+
+function open_lazy_json(path::AbstractString; kw...)
+    allocator = yyjson_alc_dyn_new()
```
Comment (gryumov) [inline src/LazyJSON.jl:271]:
```suggestion
    allocator = yyjson_alc_dyn_new()
    if allocator === C_NULL
        throw(LazyJSONError("Failed to allocate memory for JSON parsing."))
    end
```

File: src/Reader.jl [hunk 26]
```diff
@@ -26,25 +26,25 @@ function bitwise_read_flag(;
...
+function open_json_doc(path::AbstractString; alc = C_NULL, kw...)
     err = YYJSONReadErr()
     doc_ptr = yyjson_read_file(
         path,
         bitwise_read_flag(; kw...),
-        C_NULL,
+        alc,
         pointer_from_objref(err),
     )
     doc_ptr == C_NULL && throw(err)
```
Comment (gryumov) [inline src/Reader.jl]:
```suggestion
    doc_ptr === C_NULL && throw(err)
```

File: src/LazyJSON.jl [hunk 27]
```diff
@@ -0,0 +1,200 @@
...
+    elseif yyjson_is_arr(ptr)
+        LazyJSONVector(ptr)
+    else
+        nothing
+    end
+end
+
+function parse_json_string(ptr::Ptr{YYJSONVal})
+    ptr_char = yyjson_get_str(ptr)
+    ptr_char == C_NULL && throw(LazyJSONError("Error parsing string."))
```
Comment (gryumov) [inline src/LazyJSON.jl]:
```suggestion
    ptr_char === C_NULL && throw(LazyJSONError("Error parsing string."))
```

File: src/LazyJSON.jl [hunk 28]
```diff
@@ -0,0 +1,200 @@
...
+    return unsafe_string(ptr_char)
+end
+
+function parse_json_number(ptr::Ptr{YYJSONVal})
+    return yyjson_is_real(ptr) ? yyjson_get_real(ptr) : yyjson_get_int(ptr)
+end
+
+function parse_json_root(doc_ptr::Ptr{YYJSONDoc})
+    root_ptr = yyjson_doc_get_root(doc_ptr)
+    root_ptr == C_NULL && throw(LazyJSONError("Error parsing root."))
```
Comment (gryumov) [inline src/LazyJSON.jl]:
```suggestion
    root_ptr === C_NULL && throw(LazyJSONError("Error parsing root."))
```

File: src/Reader.jl [hunk 29]
```diff
@@ -26,25 +26,25 @@ function bitwise_read_flag(;
...
     err = YYJSONReadErr()
     doc_ptr = yyjson_read_opts(
         json,
         ncodeunits(json),
         bitwise_read_flag(; kw...),
-        C_NULL,
+        alc,
         pointer_from_objref(err),
     )
     doc_ptr == C_NULL && throw(err)
```
Comment (gryumov) [inline src/Reader.jl]:
```suggestion
    doc_ptr === C_NULL && throw(err)
```

File: src/LazyJSON.jl (hunk 23 above)
Comment (AlexKlo) [reply src/LazyJSON.jl:87]: My opinion is that the original looks more readable.

File: src/LazyJSON.jl [hunk 30]
```diff
@@ -0,0 +1,324 @@
...
+struct LazyJSONError <: Exception
+    message::String
+end
+
+Base.show(io::IO, e::LazyJSONError) = print(io, e.message)
+
+#__ LazyJSONDict
+
+"""
+    LazyJSONDict <: AbstractDict{AbstractString, Any}
```
Comment (artememelin) [inline src/LazyJSON.jl]:
```suggestion
    LazyJSONDict <: AbstractDict{AbstractString,Any}
```

File: src/LazyJSON.jl [hunk 31]
```diff
@@ -0,0 +1,324 @@
...
+
+function parse_json_root(doc_ptr::Ptr{YYJSONDoc})
+    root_ptr = yyjson_doc_get_root(doc_ptr)
+    root_ptr === YYJSONVal_NULL && throw(LazyJSONError("Error parsing root."))
+    return root_ptr
+end
+
+"""
+    parse_lazy_json(json::AbstractString; kw...)
+    parse_lazy_json(json::AbstractVector{UInt8}; kw...)
```
Comment (artememelin) [inline src/LazyJSON.jl]:
```suggestion
    parse_lazy_json(json::String; kw...)
    parse_lazy_json(json::Vector{UInt8}; kw...)
```

File: src/LazyJSON.jl [hunk 32]
```diff
@@ -0,0 +1,324 @@
...
+    allocator === YYJSONAlc_NULL && throw(LazyJSONError("Failed to allocate memory for JSON parsing."))
+    doc_ptr = open_json_doc(path; alc = allocator, kw...)
+    root_ptr = parse_json_root(doc_ptr)
+    root =  yyjson_is_obj(root_ptr) ? LazyJSONDict(root_ptr, doc_ptr, allocator) : LazyJSONVector(root_ptr, doc_ptr, allocator)
+    finalizer(close, root)
+    return root
+end
+
+"""
+    open_lazy_json(path::AbstractString; kw...)
```
Comment (artememelin) [inline src/LazyJSON.jl]:
```suggestion
    open_lazy_json(io::IO; kw...)
```

File: src/LazyJSON.jl [hunk 33]
```diff
@@ -0,0 +1,324 @@
...
+    root = parse_lazy_json(x...; kw...)
+    try
+        f(root)
+    finally
+        close(root)
+    end
+end
+
+"""
+    open_lazy_json(path::AbstractString; kw...)
```
Comment (artememelin) [inline src/LazyJSON.jl]:
```suggestion
    open_lazy_json(path::String; kw...)
```

File: src/LazyJSON.jl [hunk 34]
```diff
@@ -0,0 +1,324 @@
...
+Dict{Any, Any} with 2 entries:
+  "b" => nothing
+  "a" => 1
+
+julia> array
+3-element Vector{Any}:
+ 1
+ 2
+  Dict{AbstractString, Any}("b" => nothing, "a" => 3)
+```
```
Comment (artememelin) [inline src/LazyJSON.jl]: Can you make this example a little bit easy or separate into two examples?

File: src/LazyJSON.jl [hunk 35]
```diff
@@ -0,0 +1,324 @@
...
+    root_ptr === YYJSONVal_NULL && throw(LazyJSONError("Error parsing root."))
+    return root_ptr
+end
+
+"""
+    parse_lazy_json(json::AbstractString; kw...)
+    parse_lazy_json(json::AbstractVector{UInt8}; kw...)
+
+Parse a JSON string `json` (or vector of `UInt8`) into a [`LazyJSONDict`](@ref) or [`LazyJSONVector`](@ref).
+
```
Comment (artememelin) [inline src/LazyJSON.jl:182]: Add a warning about the need to close the object for reading after finishing work
And a reference to the method that takes a function as the first argument, which itself will close the lazy object

File: src/LazyJSON.jl (hunk 31 above)
Comment (AlexKlo) [reply src/LazyJSON.jl:181]: Abstract types are better, they extend the functionality

File: src/LazyJSON.jl [hunk 36]
```diff
@@ -0,0 +1,295 @@
...
+!!! note
+    When it's no longer needed, it should be closed with `close` or use `do` block syntax.
+
+## Keyword arguments
+Similar to [`parse_json`](@ref).
+
+## Examples
+```julia
```
Comment (artememelin) [inline src/LazyJSON.jl]:
```suggestion
```julia-repl
```

File: src/LazyJSON.jl [hunk 37]
```diff
@@ -0,0 +1,295 @@
...
+LazyJSONDict with 6 entries:
+  "str"     => "John Doe"
+  "num"     => "30"
+  "array"   => Any[1, 2, LazyJSONDict("a"=>3, "b"=>nothing)]
+  "bool"    => false
+  "obj"     => LazyJSONDict("a"=>1, "b"=>nothing)
+  "another" => "key"
+```
+
+```julia
```
Comment (artememelin) [inline src/LazyJSON.jl]:
```suggestion
```julia-repl
```

File: src/LazyJSON.jl [hunk 38]
```diff
@@ -0,0 +1,295 @@
...
+
+function open_lazy_json(path::AbstractString; kw...)
+    allocator = yyjson_alc_dyn_new()
+    allocator === YYJSONAlc_NULL && throw(LazyJSONError("Failed to allocate memory for JSON parsing."))
+    doc_ptr = open_json_doc(path; alc = allocator, kw...)
+    root_ptr = parse_json_root(doc_ptr)
+    root =  yyjson_is_obj(root_ptr) ? LazyJSONDict(root_ptr, doc_ptr, allocator) : LazyJSONVector(root_ptr, doc_ptr, allocator)
+    finalizer(close, root)
+    return root
+end
```
Comment (gryumov) [inline src/LazyJSON.jl]:
```suggestion
function open_lazy_json(path::AbstractString; kw...)
    allocator = yyjson_alc_dyn_new()
    allocator === YYJSONAlc_NULL && throw(LazyJSONError("Failed to allocate memory for JSON parsing."))
    doc_ptr = open_json_doc(path; alc = allocator, kw...)
    root_ptr = parse_json_root(doc_ptr)
    root = yyjson_is_obj(root_ptr) ? LazyJSONDict(root_ptr, doc_ptr, allocator) : LazyJSONVector(root_ptr, doc_ptr, allocator)
    finalizer(close, root)
    return root
end
```

File: src/LazyJSON.jl [hunk 39]
```diff
@@ -0,0 +1,295 @@
...
+    else
+        nothing
+    end
+end
+
+function parse_json_string(ptr::Ptr{YYJSONVal})
+    ptr_char = yyjson_get_str(ptr)
+    ptr_char === YYJSONUInt8_NULL && throw(LazyJSONError("Error parsing string."))
+    return unsafe_string(ptr_char)
+end
```
Comment (gryumov) [inline src/LazyJSON.jl]:
```suggestion
function parse_json_string(ptr::Ptr{YYJSONVal})
    char_ptr = yyjson_get_str(ptr)
    char_ptr === YYJSONUInt8_NULL && throw(LazyJSONError("Error parsing string."))
    return unsafe_string(char_ptr)
end
```

**Review by gryumov — APPROVED** [review]

**Pull request merged**